File: Q1046 - Last Stone Weight/1046.py
class Solution:
    def lastStoneWeight(self, stones: List[int]) -> int:

        length = len(stones)
        while length > 1:
            stones = sorted(stones)
            if stones[-1] == stones[-2]:
                stones = stones[0:length-2]
            else:
                stones[-2] = abs(stones[-1] - stones[-2])
                stones = stones[0:length-1]
            length = len(stones)

        if stones == []: return 0
        else: return stones[0]

File: Q-1046 - Last Stone Weight/1046.cpp
class Solution {
public:
    int lastStoneWeight(vector<int>& stones) {
        make_heap(stones.begin(),stones.end());
        while(true){
            if(stones.size()<=1)
                break;
            int y = stones.front();
            pop_heap(stones.begin(), stones.end());
            stones.pop_back();
            int x = stones.front();
            pop_heap(stones.begin(), stones.end());
            stones.pop_back();
            if(x<y){
                stones.push_back(y-x);
                push_heap(stones.begin(), stones.end()); 
            }
        }

        if(stones.size()==1)
            return stones[0];
        return 0;
    }
};